### PR TITLE
feat: runtime correction of whitespaces in ENV via proxy AppEnv

### PR DESCRIPTION
**Pull request by tucker-weed**
Strips whitespace off of rails environment variables before the rails app initializes.

### CODE REVIEW

File: config/environment.rb
Comment (cbarber) [inline config/environment.rb:2]: The risk here is we'll mutate all environment variables rather than just the ones our application cares about. This could have a surprise side effect if some rails variable needs leading or trailing whitespace.

What do you think of an AppEnv singleton class that removes the white space for us?
`AppEnv['GITHUB_WEBHOOK_SECRET']` returns `ENV['GITHUB_WEBHOOK_SECRET'].strip`

Comment (tucker-weed) [reply config/environment.rb:2]: I like that idea - looks like it allows for more granular control over specific environment variables (beyond just .strip as well). Will work on that now.

File: config/environment.rb
Comment (cbarber) [inline config/environment.rb:15]: Sorry I wasn't clear. I was thinking we'd avoid mutating the ENV variables directly:
```
class AppEnv
  def self.[](key)
    ENV[key].strip
  end
end
```
and then find instances of `ENV[...]` in the `app` directory to replace them with `AppEnv[...]`

I think we should also put AppEnv in its own file in `lib/app_env.rb`

File: config/environment.rb
Comment (fallynlogan) [inline config/environment.rb:4]: For this part `ENV.key?(key)` are we checking that the key is not null? Also, why would we want to include `&&` instead of just separating these onto different lines?

File: config/environment.rb
Comment (fallynlogan) [reply config/environment.rb]: @cbarber is there an example in this project where an environment variable has trailing whitespaces that are needed? If so, why is that the case?

**Comment by tmsavage** [comment]
@tucker-weed if you need help 'condensing' your 5 commits into 1, I can help you with that.

File: config/environment.rb
Comment (tmsavage) [reply config/environment.rb:15]: What instances would leading spaces be necessary for some rails variables? If we know these instances, would it be less costly to overwrite specific variables in the ENV file once so the application doesn't have to keep parsing the white spaces every time we start the app?

**Comment by sandipa12** [comment]
what does  ENV.key?(key) evaluate?

File: config/environment.rb
Comment (tucker-weed) [reply config/environment.rb]: .strip is O(n) where n is the length of the string. I think that because the length of the strings that need to be stripped are not long (around 25) and not variable in length, it is probably not a bottleneck in the runtime of the application. What are everyone's thoughts on this?

File: config/environment.rb
Comment (tucker-weed) [reply config/environment.rb]: Yes, it is a check that a key is not null (kind of functions as a defensive check at this point). The && is included so that if `ENV.key?(key)` is equal to null, the second part of the statement, `ENV[key] = ENV[key].strip` will not be evaluated. If this is not the proper way to do this, let me know.

**Comment by tucker-weed** [comment]
> what does ENV.key?(key) evaluate?

This evaluates whether a key exists in the ENV hash.

File: config/environment.rb
Comment (joealba) [reply config/environment.rb]: Looks like `n` is the length of the leading and trailing whitespace (https://ruby-doc.org/core-3.0.1/String.html#method-i-strip / https://bugs.ruby-lang.org/issues/12788), but either way isn't a problem here.  If you're concerned about `String#strip` performance, you're already using the wrong language.  ;)

File: app/lib/app_env.rb
Comment (joealba) [inline app/lib/app_env.rb:1]: I'm not sure the comment adds much value here due to the simplicity of the code.

**Comment by joealba** [comment]
I recommend correcting the commit string before merging.  Whitespace is being stripped at runtime during each access to a value now, not during the init phase.  Otherwise, 👍 💯 !

File: config/environment.rb
Comment (chewnoill) [reply config/environment.rb]: I don't think there are examples in **any** project. Wanting to keep leading/trailing whitespace is most likely a non-issue. Just impolite, maybe, to edit env variables for the whole system. Adding an abstraction on top of ENV to strip whitespace adds extra complexity to a solution that is generally just "update your env variables to the correct value". 

The real problem here, I think, is that rails fails to load, or fails to provide some critical function, because env variables are not set correctly and doesn't provide helpful information to the operator that an action is required. Stripping whitespace that was accidentally copy and pasted in solves only a symptom of this issue.

File: config/environment.rb
Comment (cbarber) [reply config/environment.rb]: I don't think we should mutate the existing ENV hash since it would be difficult to discover and troubleshoot *if* it is the ever the wrong operation.

I would instead opt for adding a new class:
```
class AppEnv
  def self.[](key)
    ENV[key].strip
  end
end
```
and changing existing usages of `ENV` found in our `app/` folder. An example:
`app/controllers/web_hooks/github_controller.rb`
```
 35   private def webhook_secret(_payload)
 36     ENV['GITHUB_WEBHOOK_SECRET']
 37   end
```
would become
```
 35   private def webhook_secret(_payload)
 36     AppEnv['GITHUB_WEBHOOK_SECRET']
 37   end
 ```

Why this change instead of the original?
* Set up other developers to do the right thing. Some time in the future when another developer wants to add a new ENV variable for some new service, they're probably going to look in the application for other services that do the same thing and copy it. With `AppEnv` used in our app for all of our env vars, they will likely copy that. They might miss the allow/disallow list of env variables to mutate in ENV.
* Discoverability. If someone needs to find out why whitespace is being stripped out, `AppEnv` in their call provides a clear callstack to follow. Trying to find a mutation in `config/environment.rb` is harder.

I think these are good coding practices to follow, even if the likelihood of us or rails ever needing whitespace in our ENV vars is pretty low.

File: app/lib/app_env.rb
Comment (tucker-weed) [reply app/lib/app_env.rb]: Removed the comment - I completely agree that it is an unnecessary annotation of the class.

File: config/environment.rb
Comment (tucker-weed) [reply config/environment.rb]: This code snippet is slightly outdated though - `ENV[key] = ENV[key].strip` is actually just `ENV[key].strip` so that the stripped whitespace ENV value is returned from the method call rather than mutating ENV permanently.

**Review by chewnoill — APPROVED** [review]
🚀

**Pull request merged**
